### ISSUE TITLE
Formatting optional parentheses around single constraints

### ISSUE THREAD

**Issue by Soupstraw** [fourmolu/fourmolu#288]
**Describe the option you'd like to add**
It would be great to add a boolean option that toggles the removal of parentheses around single constraints.
* `removeSingleConstraintParentheses` `true`/`false`
* What option setting would keep current behavior? `false`

**Show a code example exemplifying the different option settings**
```hs
-- Before formatting
f :: (Show a) => a -> String
f x = "foo" ++ show x

-- After formatting with the option enabled
f :: Show a => a -> String
f x = "foo" ++ show x
```

**Comment by Soupstraw** [comment]
Ok, done: tweag/ormolu#967

**Comment by georgefst** [comment]
I expect they're unlikely to act on that one, with it having been done intentionally for reasons of consistency and smaller diffs. It should be easy to implement in Fourmolu, however.

**Comment by Soupstraw** [comment]
Ok, so there's another issue about that already and there people are arguing for enforcing parentheses even around single constraints. Maybe this option should have three possibilities: `ignore`, `always` and `never`

**Comment by lehins** [comment]
I think this option could also apply to the derived constraints as well:

Before
```haskell
data Foo a = Foo
  deriving (Show)
```

After
```haskell
data Foo a = Foo
  deriving Show
```


Naturally, also to the superclass constraints for class definitions and instances:

Before:
```haskell
instance (Eq a) => Show (Foo a) where
  ...
```

After:
```haskell
instance Eq a => Show (Foo a) where
  ...
```

**Comment by brandonchinn178** [comment]
I agree with @lehins; I think perhaps a `single-constraint-parens: always|never|auto` option would be good (`auto` same as "ignore"; leave it how the user styled it), defaulting to `auto`

**Comment by sgillespie** [comment]
I took a stab at implementing this. Please let me know if there is anything else you need for me.

**Comment by brandonchinn178** [comment]
Fixed by #304
